// @warp-drive/build-config/dist/debugging-VdsvkNjX.js
/**
 * # Log Instrumentation <Badge type="tip" text="debug only" />
 *
 * Many portions of the internals are helpfully instrumented with logging.
 * This instrumentation is always removed from production builds.
 *
 * Log instrumentation is "regionalized" to specific concepts and concerns
 * to enable you to enable/disable just the areas you are interested in.
 *
 * To activate a particular group of logs set the appropriate flag to `true`
 * either in your build config or via the runtime helper.
 *
 *
 * ## Runtime Activation
 *
 * ::: tip 💡 Just Works in browser Dev Tools!
 * No import is needed, and the logging config is preserved when the page is refreshed
 * :::
 *
 * ```ts
 * setWarpDriveLogging({
 *   LOG_CACHE: true,
 *   LOG_REQUESTS: true,
 * })
 * ```
 *
 * A runtime helper is attached to `globalThis` to enable activation of the logs
 * from anywhere in your application including from the devtools panel.
 *
 * The runtime helper overrides any build config settings for the given flag
 * for the current browser tab. It stores the configuration you give it in
 * `sessionStorage` so that it persists across page reloads of the current tab,
 * but not across browser tabs or windows.
 *
 * If you need to deactivate the logging, you can call the helper again with the
 * same flag set to `false` or just open a new tab/window.
 *
 * ## Buildtime Activation
 *
 * ```ts
 * setConfig(__dirname, app, {
 *   debug: {
 *     LOG_CACHE: true,
 *     LOG_REQUESTS: false,
 *     LOG_NOTIFICATIONS: true,
 *   }
 * });
 * ```
 *
 * The build config settings are used to set the default values for the
 * logging flags. Any logging flag that is not set in the build config
 * will default to `false`.
 *
 * @module
 */
/**
 * log cache updates for both local
 * and remote state. Note in some older versions
 * this was called `LOG_PAYLOADS` and was one
 * of three flags that controlled logging of
 * cache updates. This is now the only flag.
 *
 * The others were `LOG_OPERATIONS` and `LOG_MUTATIONS`.
 *
 * @public
 * @since 5.5
 */
const LOG_CACHE = false;

/**
 * <Badge type="danger" text="removed" />
 *
 * This flag no longer has any effect.
 *
 * Use {@link LOG_CACHE} instead.
 *
 * @deprecated removed in version 5.5
 * @public
 */
const LOG_PAYLOADS = false;

/**
 * <Badge type="danger" text="removed" />
 *
 * This flag no longer has any effect.
 *
 * Use {@link LOG_CACHE} instead.
 *
 * @deprecated removed in version 5.5
 * @public
 */
const LOG_OPERATIONS = false;

/**
 * <Badge type="danger" text="removed" />
 *
 * This flag no longer has any effect.
 *
 * Use {@link LOG_CACHE} instead.
 *
 * @deprecated removed in version 5.5
 * @public
 */
const LOG_MUTATIONS = false;

/**
 * Log decisions made by the Basic CachePolicy
 *
 * @public
 */
const LOG_CACHE_POLICY = false;

/**
 * log notifications received by the NotificationManager
 *
 * @public
 */
const LOG_NOTIFICATIONS = false;
/**
 * log requests issued by the RequestManager
 *
 * @public
 */
const LOG_REQUESTS = false;
/**
 * log updates to requests the store has issued to
 * the network (adapter) to fulfill.
 *
 * @public
 */
const LOG_REQUEST_STATUS = false;
/**
 * log peek, generation and updates to
 * Record Identifiers.
 *

 * @public
 */
const LOG_IDENTIFIERS = false;
/**
 * log updates received by the graph (relationship pointer storage)
 *
 * @public
 */
const LOG_GRAPH = false;
/**
 * log creation/removal of RecordData and Record
 * instances.
 *
 * @public
 */
const LOG_INSTANCE_CACHE = false;
/**
 * Log key count metrics, useful for performance
 * debugging.
 *
 * @public
 */
const LOG_METRIC_COUNTS = false;
/**
 * Helps when debugging causes of a change notification
 * when processing an update to a hasMany relationship.
 *
 * @public
 */
const DEBUG_RELATIONSHIP_NOTIFICATIONS = false;

/**
 * A private flag to enable logging of the native Map/Set
 * constructor and method calls.
 *
 * EXTREMELY MALPERFORMANT
 *
 * LOG_METRIC_COUNTS must also be enabled.
 *
 * @internal
 */
const __INTERNAL_LOG_NATIVE_MAP_SET_COUNTS = false;

const LOGGING = /*#__PURE__*/Object.freeze(/*#__PURE__*/Object.defineProperty({
	__proto__: null,
	DEBUG_RELATIONSHIP_NOTIFICATIONS,
	LOG_CACHE,
	LOG_CACHE_POLICY,
	LOG_GRAPH,
	LOG_IDENTIFIERS,
	LOG_INSTANCE_CACHE,
	LOG_METRIC_COUNTS,
	LOG_MUTATIONS,
	LOG_NOTIFICATIONS,
	LOG_OPERATIONS,
	LOG_PAYLOADS,
	LOG_REQUESTS,
	LOG_REQUEST_STATUS,
	__INTERNAL_LOG_NATIVE_MAP_SET_COUNTS
}, Symbol.toStringTag, { value: 'Module' }));

export { DEBUG_RELATIONSHIP_NOTIFICATIONS as D, LOGGING as L, __INTERNAL_LOG_NATIVE_MAP_SET_COUNTS as _, LOG_CACHE as a, LOG_PAYLOADS as b, LOG_OPERATIONS as c, LOG_MUTATIONS as d, LOG_CACHE_POLICY as e, LOG_NOTIFICATIONS as f, LOG_REQUESTS as g, LOG_REQUEST_STATUS as h, LOG_IDENTIFIERS as i, LOG_GRAPH as j, LOG_INSTANCE_CACHE as k, LOG_METRIC_COUNTS as l };
